File: bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/widgets/Tree.java
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

package org.eclipse.rap.rwt.widgets;

import java.util.ArrayList;
import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.events.*;
import org.eclipse.rap.rwt.internal.widgets.*;
import org.eclipse.rap.rwt.internal.widgets.WidgetTreeVisitor.AllWidgetTreeVisitor;

/**
 * TODO: [fappel] comment
 */
public class Tree extends Composite {

  private static final TreeItem[] EMPTY_SELECTION = new TreeItem[ 0 ];
  
  private final ItemHolder itemHolder;
  private TreeItem[] selection;


  public Tree( final Composite parent, final int style ) {
    super( parent, checkStyle( style ) );
    itemHolder = new ItemHolder( TreeItem.class );
    selection = EMPTY_SELECTION;
  }

  public Object getAdapter( final Class adapter ) {
    Object result;
    if( adapter == IItemHolderAdapter.class ) {
      result = itemHolder;
    } else {
      result = super.getAdapter( adapter );
    }
    return result;
  }
  
  ///////////////////////////
  // Methods to manage items 

  public int getItemCount() {
    checkWidget();
    return itemHolder.size();
  }

  public TreeItem[] getItems() {
    checkWidget();
    return ( TreeItem[] )itemHolder.getItems();
  }
  
  public TreeItem getItem( final int index ) {
    checkWidget();
    return ( TreeItem )itemHolder.getItem( index );
  }
  
  public int indexOf( final TreeItem item ) {
    checkWidget();
    if( item == null ) {
      RWT.error( RWT.ERROR_NULL_ARGUMENT );
    }
    if( item.isDisposed() ) {
      RWT.error( RWT.ERROR_INVALID_ARGUMENT );
    }
    return itemHolder.indexOf( item );
  }
  
  public void removeAll() {
    checkWidget();
    TreeItem[] items = getItems();
    for( int i = 0; i < items.length; i++ ) {
      items[ i ].dispose();
    }
    selection = EMPTY_SELECTION;
  }
  
  /////////////////////////////////////
  // Methods to get/set/clear selection
  
  public TreeItem[] getSelection() {
    checkWidget();
    TreeItem[] result = new TreeItem[ selection.length ];
    System.arraycopy( selection, 0, result, 0, selection.length );
    return result;
  }
  
  public int getSelectionCount() {
    checkWidget();
    return selection.length;
  }
  
  public void setSelection( final TreeItem selection ) {
    checkWidget();
    if( selection == null ) {
      RWT.error( RWT.ERROR_NULL_ARGUMENT );
    }
    setSelection( new TreeItem[] { selection } );
  }

  public void setSelection( final TreeItem[] selection ) {
    checkWidget();
    if( selection == null ) {
      RWT.error( RWT.ERROR_NULL_ARGUMENT );
    }
    int length = selection.length;
    if( ( style & RWT.SINGLE ) != 0 ) {
      if( length == 0 || length > 1 ) {
        deselectAll();
      } else {
        TreeItem item = selection[ 0 ];
        if( item != null ) {
          if( item.isDisposed() ) {
            RWT.error( RWT.ERROR_INVALID_ARGUMENT );
          }
          this.selection = new TreeItem[] { item };
        }
      }
    } else {
      if( length == 0 ) {
        deselectAll();
      } else {
        // Construct an array that contains all non-null items to be selected
        TreeItem[] validSelection = new TreeItem[ length ];
        int validLength = 0;
        for( int i = 0; i < length; i++ ) {
          if( selection[ i ] != null ) {
            if( selection[ i ].isDisposed() ) {
              RWT.error( RWT.ERROR_INVALID_ARGUMENT );
            }
            validSelection[ validLength ] = selection[ i ];
            validLength++;
          }
        }
        if( validLength > 0 ) {
          // Copy the above created array to its 'final destination'
          this.selection = new TreeItem[ validLength ];
          System.arraycopy( validSelection, 0, this.selection, 0, validLength );
        }
      }
    }
  }
  
  public void selectAll() {
    checkWidget();
    if( ( style & RWT.MULTI ) != 0 ) {
      final java.util.List allItems = new ArrayList();
      WidgetTreeVisitor.accept( this, new AllWidgetTreeVisitor() {
        public boolean doVisit( final Widget widget ) {
          if( widget instanceof TreeItem ) {
            allItems.add( widget );
          }
          return true;
        }
      } );
      selection = new TreeItem[ allItems.size() ];
      allItems.toArray( selection );
    }
  }
  
  public void deselectAll() {
    checkWidget();
    this.selection = EMPTY_SELECTION;
  }

  //////////////////////////////////////
  // Listener registration/deregistration
  
  public void addSelectionListener( final SelectionListener listener ) {
    checkWidget();
    SelectionEvent.addListener( this, listener );
  }

  public void removeSelectionListener( final SelectionListener listener ) {
    checkWidget();
    SelectionEvent.removeListener( this, listener );
  }

  public void addTreeListener( final TreeListener listener ) {
    checkWidget();
    TreeEvent.addListener( this, listener );
  }
  
  public void removeTreeListener( final TreeListener listener ) {
    checkWidget();
    TreeEvent.removeListener( this, listener );
  }

  ////////////////////////////////
  // Methods to cleanup on dispose
  
  protected void releaseChildren() {
    TreeItem[] items = getItems();
    for( int i = 0; i < items.length; i++ ) {
      items[ i ].dispose();
    }
    super.releaseChildren();
  }
  
  void removeFromSelection( final TreeItem item ) {
    int index = -1;
    for( int i = 0; index == -1 && i < selection.length; i++ ) {
      if( selection[ i ] == item ) {
        index = i;
      }
    }
    if( index != -1 ) {
      TreeItem[] newSelection = new TreeItem[ selection.length - 1 ];
      System.arraycopy( selection, 0, newSelection, 0, index );
      if( index < selection.length - 1 ) {
        int length = selection.length - index - 1;
        System.arraycopy( selection, index + 1, newSelection, index, length );
      }
      selection = newSelection;
    }
  }
  
  //////////////////
  // Helping methods
  
  private static int checkStyle( final int style ) {
    int result = style | RWT.H_SCROLL | RWT.V_SCROLL;
    return checkBits( result, RWT.SINGLE, RWT.MULTI, 0, 0, 0, 0 );
  }
}

File: bundles/org.eclipse.rap.rwt/js/org/eclipse/rap/rwt/widgets/Tree.js
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

/**
 * This class encapulates the qx.ui.treefullcontrol.Tree o make it more
 * suitable for usage in RWT.
 * The style parameter mimics the RWT style flag. Possible values (strings)
 * are: multi, check
 */
qx.OO.defineClass(
  "org.eclipse.rap.rwt.widgets.Tree",
  qx.ui.treefullcontrol.Tree,
  function( style ) {
    var trs = qx.ui.treefullcontrol.TreeRowStructure.getInstance().standard( "" );
    qx.ui.treefullcontrol.Tree.call( this, trs );
    this.setOverflow( qx.constant.Style.OVERFLOW_AUTO );
    this.setHideNode( true );
    this.setUseTreeLines( true );
    this.setUseDoubleClick( false ); // true supresses dblclick events !
    // TODO [rh] this is only to make the tree fousable at all
    this.setTabIndex( 1 );
    this._rwtStyle = style;
    this._selectionListeners = false;
    this._treeListeners = false;
    var manager = this.getManager();
    manager.setMultiSelection( qx.lang.String.contains( style, "multi" ) );
    manager.addEventListener( "changeSelection", this._onChangeSelection, this );
    this.addEventListener( "treeOpenWithContent", this._onItemExpanded, this );
    this.addEventListener( "treeClose", this._onItemCollapsed, this );
    this.addEventListener( "contextmenu", this._onContextMenu, this );
    this.addEventListener( "changeEnabled", this._onChangeEnabled, this );
  }
);

/**
 * Are there any server-side SelectionListeners attached? If so, selecting an
 * item causes a request to be sent that informs the server-side listeners.
 */
qx.Proto.setSelectionListeners = function( value ) {
  this._selectionListeners = value;
}

qx.Proto.hasSelectionListeners = function() {
  return this._selectionListeners;
}

/**
 * Are there any server-side TreeListeners attached? If so, expanding/collapsing
 * an item causes a request to be sent that informs the server-side listeners.
 */
qx.Proto.setTreeListeners = function( value ) {
  this._treeListeners = value;  
}

qx.Proto.getRWTStyle = function() {
  return this._rwtStyle;  
}

qx.Proto.dispose = function() {
  if( this.getDisposed() ) {
    return true;
  }
  var manager = this.getManager();
  manager.removeEventListener( "changeSelection", this._onChangeSelection, this );
  this.removeEventListener( "treeOpenWithContent", this._onItemExpanded, this );
  this.removeEventListener( "treeClose", this._onItemCollapsed, this );
  this.removeEventListener( "changeEnabled", this._onChangeEnabled, this );
  return qx.ui.treefullcontrol.Tree.prototype.dispose.call( this );
}

//
// Event Listener
//

qx.Proto._onChangeSelection = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
    var req = org.eclipse.rap.rwt.Request.getInstance();
    var id = wm.findIdByWidget( this );
    var item = this.getManager().getLeadItem();
    req.addParameter( id + ".selection", this._getSelectionIndices() );
    // TODO [rst] Prevent selecting the root item.
    //      When first visible item is selected and arrow up is pressed the root
    //      item ( == this ) is selected which results in an invisible selection. 
    if( item == this ) {
//      this.getFirstVisibleChildOfFolder().setSelected( true );
//      this.setSelected( false );
    } else {
      if( this._selectionListeners ) {
        this._suspendClicks();
        var itemId = wm.findIdByWidget( item );
        var eventName = "org.eclipse.rap.rwt.events.widgetSelected";
        req.addEvent( eventName, id );
        req.addParameter( eventName + ".item", itemId );
        req.send();
      }
    }
  }
}

qx.Proto._onItemExpanded = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
    var treeItemId = wm.findIdByWidget( evt.getData() );
    var req = org.eclipse.rap.rwt.Request.getInstance();
    req.addParameter( treeItemId + ".state", "expanded" );
    if( this._treeListeners ) {
      req.addEvent( "org.eclipse.rap.rwt.events.treeExpanded", treeItemId );
      req.send();
    }
  }
}

qx.Proto._onItemCollapsed = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
    var treeItemId = wm.findIdByWidget( evt.getData() );
    var req = org.eclipse.rap.rwt.Request.getInstance();
    req.addParameter( treeItemId + ".state", "collapsed" );
    if( this._treeListeners ) {
      req.addEvent( "org.eclipse.rap.rwt.events.treeCollapsed", treeItemId );
      req.send();
    }
  }
}

qx.Proto._onContextMenu = function( evt ) {
  var menu = this.getContextMenu();
  if( menu != null ) {
    menu.setLocation( evt.getPageX(), evt.getPageY() );
    menu.setOpener( this );
    menu.show();
    evt.stopPropagation();
  }
}

/*
 * Pass enablement to tree items
 */
qx.Proto._onChangeEnabled = function( evt ) {
  var newValue = evt.getData();
  var items = this.getItems();
  for( var i = 0; i < items.length; i++ ) {
    var item = items[ i ];
    if( item.getLabelObject() != null ) {
      var label = item.getLabelObject();
      label.setEnabled( newValue );
    } else {
      // TODO [rh] revise this: how to remove/dispose of the listener?
      item.addEventListener( "appear", function( evt ) {
        this.getLabelObject().setEnabled( newValue );
      }, item );
    }
    item.setEnabled( newValue );
  }
}

/*
 * handle click on tree item
 * called by org.eclipse.rap.rwt.widgets.TreeItem
 */
qx.Proto._notifyItemClick = function( item ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    if( this._selectionListeners && !this._clicksSuspended ) {
      this._suspendClicks();
      var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
      var id = wm.findIdByWidget( this );
      var itemId = wm.findIdByWidget( item );
      var req = org.eclipse.rap.rwt.Request.getInstance();
      var eventName = "org.eclipse.rap.rwt.events.widgetSelected";
      req.addEvent( eventName, id );
      req.addParameter( eventName + ".item", itemId );
      req.send();
    }
  }
}

/*
 * handle double click on tree item
 * called by org.eclipse.rap.rwt.widgets.TreeItem
 */
qx.Proto._notifyItemDblClick = function( item ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    if( this._selectionListeners ) {
      var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
      var id = wm.findIdByWidget( this );
      var itemId = wm.findIdByWidget( item );
      var req = org.eclipse.rap.rwt.Request.getInstance();
      var eventName = "org.eclipse.rap.rwt.events.widgetDefaultSelected";
      req.addEvent( eventName, id );
      req.addParameter( eventName + ".item", itemId );
      req.send();
    }
  }
}

/*
 * handle change of the check state of a tree item's check box
 * called by org.eclipse.rap.rwt.widgets.TreeItem
 */
qx.Proto._notifyChangeItemCheck = function( item ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend ) {
    if( this._selectionListeners ) {
      var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
      var id = wm.findIdByWidget( this );
      var itemId = wm.findIdByWidget( item );
      var req = org.eclipse.rap.rwt.Request.getInstance();
      var eventName = "org.eclipse.rap.rwt.events.widgetSelected";
      req.addEvent( eventName, id );
      req.addParameter( eventName + ".item", itemId );
      req.addParameter( eventName + ".detail", "check" );
      req.send();
    }
  }
}

/*
 * Returns the current selection as comma separated string
 */
// TODO [rh] handle multi selection
qx.Proto._getSelectionIndices = function() {
  var wm = org.eclipse.rap.rwt.WidgetManager.getInstance();
  var selectedItemIds = "";
  if( this.getManager().getMultiSelection() ) {
    var selectedItems = this.getManager().getSelectedItems();
    for( var i = 0; i < selectedItems.length; i++ ) {
      if( i > 0 ) {
        selectedItemIds += ",";
      }
      selectedItemIds += wm.findIdByWidget( selectedItems[ i ] );
    }
  } else {
    selectedItemIds = wm.findIdByWidget( this.getManager().getSelectedItem() );
  }
  return selectedItemIds;
}

/*
 * Suspends the processing of click events to avoid sending multiple
 * widgetSelected events to the server.
 */
qx.Proto._suspendClicks = function() {
  this._clicksSuspended = true;
  qx.client.Timer.once( this._enableClicks, this, 500 );
}

qx.Proto._enableClicks = function() {
  this._clicksSuspended = false;
}
